File: app/src/main/java/com/zuk0/gaijinsmash/riderz/data/local/room/dao/FavoriteDao.java
package com.zuk0.gaijinsmash.riderz.data.local.room.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import io.reactivex.Maybe;

import com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite;

import java.util.List;

@Dao
public interface FavoriteDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void save(Favorite favorite);

    @Delete
    void delete(Favorite favorite);

    @Query("SELECT * from favorites where id = :id")
    Favorite getFavoriteById(int id);

    @Query("SELECT * from favorites")
    LiveData<List<Favorite>> getAllFavoritesLiveData();

    @Query("SELECT COUNT(*) from favorites")
    int countFavorites();

    @Query("SELECT * from favorites where origin = :origin and destination = :destination")
    Favorite getFavorite(String origin, String destination);

    @Query("SELECT * from favorites where origin = :origin and destination = :destination " +
            "OR origin = :destination and destination = :origin")
    boolean doesFavoriteExist(String origin, String destination);

    @Query("SELECT * from favorites where priority = 1")
    Maybe<Favorite> getPriorityFavorite();

    //Update a priority favorite
    @Query("Update favorites set priority = 1 where id = :id")
    void updatePriorityById(int id);

    //Update priority to null
    @Query("Update favorites set priority = 0 where id = :id")
    void removePriorityById(int id);

    @Query("SELECT * from favorites where origin = :origin and destination = :destination OR origin = :destination and destination = :origin")
    LiveData<Favorite> getLiveDataFavorite(String origin, String  destination);

    @Query("SELECT count(*) from favorites where priority = 1")
    int getPriorityCount();


}

File: app/src/main/java/com/zuk0/gaijinsmash/riderz/data/local/room/dao/FavoriteDao.kt
package com.zuk0.gaijinsmash.riderz.data.local.room.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.zuk0.gaijinsmash.riderz.data.local.entity.Favorite
import com.zuk0.gaijinsmash.riderz.data.local.entity.station_response.Station
import io.reactivex.Maybe

@Dao
interface FavoriteDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun save(favorite: Favorite?)

    @Delete
    fun delete(favorite: Favorite?)

    @Query("SELECT * from favorites where id = :id")
    fun getFavoriteById(id: Int): Favorite?

    @get:Query("SELECT * from favorites")
    val allFavoritesLiveData: LiveData<MutableList<Favorite>>

    @get:Query("SELECT * from favorites")
    val list: MutableList<Favorite>?

    @Query("SELECT COUNT(*) from favorites")
    fun countFavorites(): Int

    @Query("SELECT * from favorites where a = :a and b = :b")
    fun getFavorite(a: Station?, b: Station?): Favorite?

    @get:Query("SELECT * from favorites where priority = 1")
    val priorityFavorite: Maybe<Favorite>

    //Update a priority favorite
    @Query("Update favorites set priority = 1 where id = :id")
    fun updatePriorityById(id: Int)

    //Update priority to null
    @Query("Update favorites set priority = 0 where id = :id")
    fun removePriorityById(id: Int)

    @Query("SELECT * from favorites where a = :a and b = :b OR a = :b and b = :a")
    fun getLiveDataFavorite(a: Station, b: Station): LiveData<Favorite>

    @get:Query("SELECT count(*) from favorites where priority = 1")
    val priorityCount: Int
}